lib/ocpp/v201/profile.cpp

```diff
@@ -529,7 +529,62 @@ void update_itt(const int schedule_duration, std::vector<ChargingSchedulePeriod>
-CompositeSchedule calculate_composite_schedule(const CompositeSchedule& charging_station_max,
+/// \brief update the period based on the power limit
+/// \param[in] current the current startPeriod based on duration
+/// \param[inout] period the details of the current period in the schedule.
+/// \param[in] charging_schedule_period the period that is being compared to period.
+/// \param[in] current_charging_rate_unit_enum details of the current composite schedule charging_rate_unit for
+/// conversion.
+///
+ChargingSchedulePeriod adjust_charging_schedule_period(const int current, const ChargingSchedulePeriod period,
+                                                       const ChargingSchedulePeriod charging_schedule_period,
+                                                       const ChargingRateUnitEnum current_charging_rate_unit_enum) {
+    auto _period = period;
+
+    if (charging_schedule_period.startPeriod != NO_START_PERIOD) {
+        _period.startPeriod = current;
+
+        if (period.limit == NO_LIMIT_SPECIFIED) {
+            // use ChargePointMaxProfile when TxProfile and TxDefaultProfile don't exist
+            if (charging_schedule_period.limit != NO_LIMIT_SPECIFIED) {
+                _period = charging_schedule_period;
+            }
+        } else {
+            // apply ChargePointMaxProfile limits
+            // Note the actual ChargePointMaxProfile limit is controlled outside of libocpp
+
+            if (charging_schedule_period.limit != NO_LIMIT_SPECIFIED) {
+
+                const auto charge_point_max_phases =
+                    charging_schedule_period.numberPhases.value_or(DEFAULT_AND_MAX_NUMBER_PHASES);
+                const auto period_max_phases = period.numberPhases.value_or(DEFAULT_AND_MAX_NUMBER_PHASES);
+                _period.numberPhases = std::min(charge_point_max_phases, period_max_phases);
+
+                if (current_charging_rate_unit_enum == ChargingRateUnitEnum::A) {
+                    // limit is per phase
+                    if (charging_schedule_period.limit < period.limit) {
+                        // apply lower limit
+                        _period.limit = charging_schedule_period.limit;
+                    }
+                } else {
+                    // limit is total allowed power, changes in number of phases matter
+
+                    // adjust limits taking into account a potential change in number of phases
+                    const auto charge_point_limit_per_phase = charging_schedule_period.limit / charge_point_max_phases;
+                    const auto period_limit_per_phase = period.limit / period_max_phases;
+
+                    // avoid fractional results - 1 decimal place is allowed but tricky to ensure with a float
+                    _period.limit = std::floor(std::min(charge_point_limit_per_phase, period_limit_per_phase) *
+                                               period.numberPhases.value());
+                }
+            }
+        }
+    }
```

Comment:
If everything is within this block starting from `if (charging_schedule_period.startPeriod != NO_START_PERIOD)`, why not keep that within `calculate_composite_schedule()`?